Prepare for a 0.13.0 release

joshuagl opened the pull request:
**Fixes issue #**: N/A

**Description of the changes being introduced by the pull request**:
* Add a document describing the release process
* prepare docs/CHANGELOG.md, setup.cfg and tuf/__init__.py for a 0.13.0 release

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

joshuagl committed:
Document release process steps

Write down the expected steps for a maintainer to follow when making a
release of tuf

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Prepare 0.13.0 release

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented on docs/RELEASE.md:
```diff
@@ -0,0 +1,28 @@
+  * create source dist `python setup.py sdist`
+  * create wheel (with 2 and 3 support) `python setup.py bdist_wheel --universal`
+  * Sign the dists `gpg --detach-sign -a dist/tuf-vA.B.C.tar.gz`
+  * Upload to test PyPI `twine upload --repository testpypi dist/*`
```
Can't currently perform this step, as @vladimir-v-diaz is the owner of that package on test.pypi.org

lukpueh commented on docs/RELEASE.md:
```diff
@@ -0,0 +1,28 @@
+
+* Ensure docs/CHANGELOG.md contains a one-line summary of each significant
+  change since the prior release
+* Update setup.py and tuf/__init__.py to the new version number vA.B.C
```
Very minor nit: Could you escape the double underscore in `__init__.py` (or use backticks)?

lukpueh commented on docs/RELEASE.md:
```diff
@@ -0,0 +1,28 @@
+* Test packaging, uploading to Test PyPI and installing from a virtual environment
+  * remove existing dist build dirs
+  * create source dist `python setup.py sdist`
+  * create wheel (with 2 and 3 support) `python setup.py bdist_wheel --universal`
```
Very very minor (consistency) nit:
```suggestion
  * Remove existing dist build dirs
  * Create source dist `python setup.py sdist`
  * Create wheel (with 2 and 3 support) `python setup.py bdist_wheel --universal`
```

lukpueh commented on docs/RELEASE.md:
```diff
@@ -0,0 +1,28 @@
+* Create a package for the formal release
+  * remove existing dist build dirs
+  * create source dist `python setup.py sdist`
+  * create wheel (with 2 and 3 support) `python setup.py bdist_wheel --universal`
```
Very very minor (consistency) nit:
```suggestion
  * Remove existing dist build dirs
  * Create source dist `python setup.py sdist`
  * Create wheel (with 2 and 3 support) `python setup.py bdist_wheel --universal`
```

lukpueh commented on docs/RELEASE.md:
```diff
@@ -0,0 +1,28 @@
+  * remove existing dist build dirs
+  * create source dist `python setup.py sdist`
+  * create wheel (with 2 and 3 support) `python setup.py bdist_wheel --universal`
+  * Sign the dists `gpg --detach-sign -a dist/tuf-vA.B.C.tar.gz`
```
Could we make this two lines?  E.g.
```suggestion
  * Sign source dist `gpg --detach-sign -a dist/tuf-vA.B.C.tar.gz`
  * Sign wheel `gpg --detach-sign -a dist/tuf-vA.B.C-py2.py3-none-any.whl`
```

lukpueh commented on docs/RELEASE.md:
```diff
@@ -0,0 +1,28 @@
+* Create a PR with updated CHANGELOG.md and version bumps
+* Once the PR is merged, pull the updated `develop` branch locally
+* Create a signed tag matching the updated version number on the merge commit
+  `git tag vA.B.C`
```
```suggestion
  `git tag --sign vA.B.C`
```

joshuagl replied:
Whoops, thanks for catching that!

lukpueh reviewed (changes requested):
Many thanks for preparing this release, and big kudos for adding a `RELEASE.md` document, @joshuagl. ❤️ 
The former definitely looks good to me. I double-checked that all PR numbers point to the right PR and no important PRs were omitted. You only left out dependabot PRs and a few minor changes that probably could be added to the changelog but don't need to, i.e. #1021 #1028 #1044 #1051 #1072 #1086. Maybe double-check if you left them out on purpose or not.

In RELEASE.md I left a few comments, feel free to address as you see fit. Thanks!

trishankatdatadog commented:
> Many thanks for preparing this release, and big kudos for adding a `RELEASE.md` document, @joshuagl. ❤️

💯 💯 

QQ: would it be possible to formalize `RELEASE.md` with a script?

trishankatdatadog commented:
I would also probably hold off until we merge some important fixes, including #1091

joshuagl commented:
> Many thanks for preparing this release, and big kudos for adding a `RELEASE.md` document, @joshuagl. ❤️
> The former definitely looks good to me. I double-checked that all PR numbers point to the right PR and no important PRs were omitted. You only left out dependabot PRs and a few minor changes that probably could be added to the changelog but don't need to, i.e. #1021 #1028 #1044 #1051 #1072 #1086. Maybe double-check if you left them out on purpose or not.
> 
> In RELEASE.md I left a few comments, feel free to address as you see fit. Thanks!

I may have missed, #1072, but mostly those PRs were things I opted to ignore because I didn't think they were particularly interesting. I'd be happy to include the PRs you mentioned, or not, and add some additional text to `RELEASE.md` to guide future `CHANGELOG.md` editors in curating a set of changes that better matches expectations. 

Personally I prefer a changelog to include __notable__ changes, rather than all changes. However I realise even that is subjective and there are entries I added to the changelog for this release that don't meet my own criteria for notable.

I welcome thoughts on changelog curation. In the past I've followed similar principles to https://keepachangelog.com/, which would suggest editing the changelog here to group similar changes and clearly delineate new features, changed features, fixes, security issues, etc.

joshuagl commented:
> > Many thanks for preparing this release, and big kudos for adding a `RELEASE.md` document, @joshuagl. ❤️
> 
> 💯 💯
> 
> QQ: would it be possible to formalize `RELEASE.md` with a script?

Agreed, scripting some of this would be good. I'll take a look at that in future.

joshuagl commented:
> I would also probably hold off until we merge some important fixes, including #1091

I'm hoping to get into the habit of releasing more regularly. docs/ROADMAP.md suggests the project releases every 3 months. I'd like to get closer to that, so I don't think in general we should worry too much about holding a release for fixes.

#1091 specifically, though, I'd be happy to merge and include in the release

joshuagl committed:
docs/RELEASE.md: link to guidance on changelogs

https://keepachangelog.com provides good advice on curating a changelog

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
docs/CHANGELOG.md: update for v0.13.0

Categorise changes by type, per the recommendations at keepachangelog.com

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
> #1091 specifically, though, I'd be happy to merge and include in the release

I've merged this PR and included it in `docs/CHANGELOG.md`

joshuagl committed:
setup.py: add project_urls links

These additional URLs will be displayed on PyPI:
https://packaging.python.org/guides/distributing-packages-using-setuptools/#project-urls

Signed-off-by: Joshua Lock <jlock@vmware.com>

lukpueh commented:
> I may have missed, #1072, but mostly those PRs were things I opted to ignore because I didn't think they were particularly interesting. I'd be happy to include the PRs you mentioned, or not, and add some additional text to `RELEASE.md` to guide future `CHANGELOG.md` editors in curating a set of changes that better matches expectations.
> 
> Personally I prefer a changelog to include **notable** changes, rather than all changes. However I realise even that is subjective and there are entries I added to the changelog for this release that don't meet my own criteria for notable.
> 
> I welcome thoughts on changelog curation. In the past I've followed similar principles to https://keepachangelog.com/, which would suggest editing the changelog here to group similar changes and clearly delineate new features, changed features, fixes, security issues, etc.

👍  This all makes sense and aligns with what I thought. Just wanted to double-check. :) Thanks for adding a link to the changelog guide to RELEASE.md.

lukpueh reviewed (approved):
💯
